OfficeDev/office-ui-fabric-react
# DetailsHeader: Initialize "Select All" state by reading optional selection prop's state if available

#### Pull request checklist

- [x] Addresses an existing issue: Fixes #7206

#### Description of changes

This fixes an issue where the "Select All" control's state of `DetailsList` is not rendered properly when the component is re-mounted [Codepen](https://codepen.io/hardik-chaudhary/pen/dQdgWq?editors=1010). _Currently_, only the individual rows' selection state will re-render correctly. 

The root cause is that we do not check `selection.isAllSelected()` when initializing `DetailsHeader`'s state.

We have a similar check when selection changes in:

https://github.com/OfficeDev/office-ui-fabric-react/blob/fb0dbcacc588e602708947e93868112f830707c7/packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.base.tsx#L842

I've re-generated snapshots. I spent _too much_ time trying to unit test this in the AM and am deferring that work given the scope of this change and snapshot coverage.

#### Focus areas to test

1. Confirm that the existing `DetailsList` examples render their individual and "All" rows selected UI properly.
1. Wrap the `DetailsList` in a conditional render like one that is toggled by button click.
1. "Select All" with the `DetailsList` rendered.
1. Toggle the state so the control is no longer rendered.
1. Toggle the state so the `DetailsList` is re-rendered.
1. "Select All" state should be persisted. See screen-grab below.

**Before**

![before_dl_select_all_rerender](https://user-images.githubusercontent.com/706967/49830410-f0d71b80-fd45-11e8-9660-aee9a9f63e9e.gif)


**After**

![dl_select_all_rerender](https://user-images.githubusercontent.com/706967/49830041-eec08d00-fd44-11e8-9226-19a40bb07781.gif)
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/7358)
